Stratis.Bitcoin.Features.BlockStore/LoopSteps/ProcessPendingStorageStep.cs

```diff
@@ -43,132 +42,158 @@ internal ProcessPendingStorageStep(BlockStoreLoop blockStoreLoop, ILoggerFactory
-                    break;
-                }
+        /// <summary>
+        /// When the node is in IBD continuously process all the blocks in <see cref="BlockStoreLoop.PendingStorage"/> until 
+        /// a stop condition is found.
+        /// </summary>
+        private async Task<StepResult> ProcessWhenInIBD(ProcessPendingStorageContext context)
+        {
+            this.logger.LogDebug("{0}:'{1}/{2}'", nameof(context.NextChainedBlock), context.NextChainedBlock.HashBlock, context.NextChainedBlock.Height);
+
+            if (this.BlockStoreLoop.PendingStorage.Count < BlockStoreLoop.PendingStorageBatchThreshold)
+                return StepResult.Continue;
 
-                context.GetNextBlock();
+            do
+            {
+                PrepareNextBlockFromPendingStorage(context);
 
-                var shouldBreak = ShouldBreakExecution(context);
-                if (shouldBreak)
+                var canProcessNextBlock = context.CanProcessNextBlock();
+                if (canProcessNextBlock == false)
                 {
                     if (context.PendingBlockPairsToStore.Any())
-                        await this.PushBlocksToRepository(context);
+                        await PushBlocksToRepository(context);
 
                     break;
                 }
-            }
+
+                if (context.PendingStorageBatchSize > BlockStoreLoop.MaxPendingInsertBlockSize)
+                {
+                    await PushBlocksToRepository(context);
+
+                    context.PendingBlockPairsToStore.Clear();
+                    context.PendingStorageBatchSize = 0;
+                }
+
+            } while (context.CancellationToken.IsCancellationRequested == false);
 
             return StepResult.Continue;
         }
 
         /// <summary>
-        /// Store missing blocks and remove them from pending blocks and set the Store's tip to <see cref="ProcessPendingStorageContext.NextChainedBlock"/>
+        /// When the node is NOT in IBD, process and push the blocks in <see cref="BlockStoreLoop.PendingStorage"/> immediately 
+        /// to the block repository without checking batch size.
         /// </summary>
-        private async Task PushBlocksToRepository(ProcessPendingStorageContext context)
+        private async Task<StepResult> ProcessWhenNotInIBD(ProcessPendingStorageContext context)
         {
-            await this.BlockStoreLoop.BlockRepository.PutAsync(context.PendingBlockPairsToStore.First().ChainedBlock.HashBlock, context.PendingBlockPairsToStore.Select(b => b.Block).ToList());
-            this.BlockStoreLoop.SetStoreTip(context.PendingBlockPairsToStore.First().ChainedBlock);
+            this.logger.LogDebug("{0}:'{1}/{2}'", nameof(context.NextChainedBlock), context.NextChainedBlock.HashBlock, context.NextChainedBlock.Height);
+
+            do
+            {
+                PrepareNextBlockFromPendingStorage(context);
+
+                await PushBlocksToRepository(context);
```

Comment:
Danger of pushing a block out of sync 
I think we have to put the `context.CanProcessNextBlock()` here as well before we push.

We could also take the same functionality from the IBD loop they are very similar (apart for the threshold condition)
IBD is time based  so  sometimes there could be over 100 blocks while not in IBD.

Basically you could have one method for both (reuse the logic from the IBD method) and when in IBD call it only if the threshold is correct for example
```
ProcessWhenInIBD()
{
  if (this.BlockStoreLoop.PendingStorage.Count < BlockStoreLoop.PendingStorageBatchThreshold)
                return StepResult.Continue;
  
        this.ProcessWhenNotInIBD()
}
```

What do you think?
(you could then rename ProcessWhenNotInIBD to ProcessesBatch())

Reply:
OK yes, I refactored a bit.. and now see that the code is the same, apart from the max size threshold. I've amended this now.. will commit soon :D

Reply:
@dangershony I've rethought your suggestion above and actually still feel that the logic for InIBD and NotInIBD should be in separate methods. Its easier to test and read that way cause it very clearly encapsulates the logic in one method for each scenario. Its just that otherwise, you need to start introducing more "If" statements again for IBD and disposing and thresholds etc. I like to code in a way that keeps the logic simple and easily readable  (Single responsibility).

I do however agree with you that there is some slight logic duplication and have done some work to fix it, however, I feel for sake of readability and keeping logic separate, its not that bad.

Reply:
Sure there is no rule for code against code duplication when it intentional, my point was just that the code in both cases needs to be similar.